contrib/endpoints/src/api_manager/service_control/aggregated_test.cc

```diff
@@ -195,6 +231,124 @@ TEST_F(AggregatedTestWithRealClient, CheckOKTest) {
   EXPECT_EQ(stat.send_report_operations, 0);
 }
 
+class QuotaAllocationFailedTestWithRealClient : public ::testing::Test {
+ public:
+  void SetUp() {
+    service_.set_name("test_service");
+    service_.mutable_control()->set_environment(
+        "servicecontrol.googleapis.com");
+    env_.reset(new ::testing::NiceMock<MockApiManagerEnvironment>);
+    sc_lib_.reset(Aggregated::Create(service_, nullptr, env_.get(), nullptr));
+    ASSERT_TRUE((bool)(sc_lib_));
+    // This is the call actually creating the client.
+    sc_lib_->Init();
+  }
+
+  void DoRunHTTPRequest(HTTPRequest* request) {
+    std::map<std::string, std::string> headers;
+    AllocateQuotaResponse allocate_quota_response_;
+
+    ASSERT_TRUE(::google::protobuf::TextFormat::ParseFromString(
+        kAllocateQuotaResponseErrorExhausted, &allocate_quota_response_));
+
+    std::string body = allocate_quota_response_.SerializeAsString();
+
+    request->OnComplete(Status::OK, std::move(headers), std::move(body));
+  }
+
+  ::google::api::Service service_;
+  std::unique_ptr<MockApiManagerEnvironment> env_;
+  std::unique_ptr<Interface> sc_lib_;
+};
+
+TEST_F(QuotaAllocationFailedTestWithRealClient, AllocateQuotaTest) {
+  EXPECT_CALL(*env_, DoRunHTTPRequest(_))
+      .WillOnce(Invoke(
+          this, &QuotaAllocationFailedTestWithRealClient::DoRunHTTPRequest));
+
+  std::vector<std::pair<std::string, int>> metric_cost_vector;
+  metric_cost_vector.push_back(
+      std::make_pair<std::string, int>("metric_first", 1));
+  metric_cost_vector.push_back(
+      std::make_pair<std::string, int>("metric_second", 2));
+
+  QuotaRequestInfo info;
+
+  info.metric_cost_vector = &metric_cost_vector;
+
+  FillOperationInfo(&info);
+  sc_lib_->Quota(info, nullptr, [](Status status) {
+    ASSERT_TRUE(status.code() == Code::RESOURCE_EXHAUSTED);
+  });
+}
+
+class QuotaAllocationTestWithRealClient : public ::testing::Test {
+ public:
```

Comment:
It seem your two test classes are almost the same except DoRunHTTPRequest() function.  You can combine them into one test class with two different DoRun functions.   EXPECT_CALL() can invoke different one.